src/platform/forms-system/src/js/fields/FileField.jsx

```diff
@@ -29,25 +41,122 @@ export default class FileField extends React.Component {
+    focusElement(`#${this.props.idSchema.$id}_add_label span`);
+  };
+
+  /**
+   * @typedef Files~state - File state object
+   * @type {object}
+   * @property {object|null} file - upload API supplied File object: see
+   *   https://developer.mozilla.org/en-US/docs/Web/API/File
+   * @property {number} index - index matching formData array position
+   * @property {boolean} isEncrypted - detected encrypted state of a PDF file
+   * @property {boolean} isTemporary - indicates state data that needs needs a
+   *   placeholder
+   */
+  /**
+   * Update file state
+   * @param {object|null} file - API supplied File object; store the entire File
+   *   object because it needs to be passed to the `uploadRequest` function
+   * @param {number} index - index of file in formData
+   * @param {boolean} isEncrypted - encrypted state
+   */
+  setFileState = ({ file, index, isEncrypted }) => {
+    // remove already uploaded files from state
+    const files = this.state.files.filter(
+      fileObj =>
+        this.props.formData.findIndex(
+          fileData => fileData.name === fileObj.file.name,
+        ) < 0,
+    );
+
+    const existingIndex = files.findIndex(fileData => fileData.index === index);
+    const data = { file, index, isEncrypted, isTemporary: true };
+    if (existingIndex > -1) {
+      files[existingIndex] = data;
+      this.setState({ files });
+    } else {
+      this.setState({ files: [...files, data] });
+    }
+  };
+
+  onSubmitPassword = (file, index, password) => {
+    if (file && password) {
+      this.onAddFile({ target: { files: [file] } }, index, password);
+    }
+  };
+
+  processFile = async (file, index) =>
+    checkForEncryptedPdf(file)
+      .then(result => {
+        if (result) {
+          this.setFileState({ file, index, isEncrypted: true });
+        }
+        return result;
+      })
+      .catch(() => {
+        // This _should_ only be called if a file is deleted after the
+        // user selects it for upload
+        this.setFileState({ index, isEncrypted: false });
+        return false;
+      });
+
+  /**
+   * Add file to list and upload
+   * @param {Event} event - DOM File upload event
+   * @param {number} index - uploaded file index, if already uploaded
+   * @param {string} password - encrypted PDF password, only defined by
+   *   `onSubmitPassword` function
+   * @listens
+   */
+  onAddFile = async (event, index = null, password) => {
     if (event.target.files && event.target.files.length) {
       const files = this.props.formData || [];
       let idx = index;
       if (idx === null) {
         idx = files.length === 0 ? 0 : files.length;
       }
+      const currentFile = event.target.files[0];
+
+      // Check if the file is an encrypted PDF
+      if (
+        this.state.pdfPasswordFeature &&
+        !password &&
+        currentFile.name.endsWith('pdf') &&
+        (this.props.uiSchema['ui:options'] || {}).getEncryptedPassword
+      ) {
```

Comment:
Can't a lot of this be done with a call to `checkForEncryptedPdf`?

Reply:
Maybe? These checks are done immediately after the user chooses a file to upload, so checking for a feature flag, existing password, correct file extension and uiSchema setting is much less expensive than calling a Promise. Or maybe I'm wrong? I've refactored this code so many times, I wouldn't be surprised if there was a more efficient method I overlooked.